etc/ci-prepare.sh

```diff
@@ -40,8 +40,22 @@ then
 fi
 
 
-# configure and make GAP
+# configure
 time "$SRCDIR/configure" --enable-Werror $CONFIGFLAGS
+
+if [[ $HPCGAP = yes ]]
+then
+  git clone https://github.com/rbehrends/unward
+  cd unward
+  ./configure CC=gcc CXX=g++ CFLAGS=-O2 CXXFLAGS=-O2
+  make
+  cd ..
+  unward/bin/unward --inplace src
+  # commit the result to prevent the docomp test from triggering
+  git commit -m "Unward" src
```

Comment:
So this means people who want to use the code in this PR still need to have unward, and run it, and then have lots of modifications in their tree. That's IMHO *the* major blocker for merging this PR, and until we resolve it, there is no way this can be merged. So I see two ways forward:

1. We run `unward` and commit the modifications once and for all, and move on (my understanding always was that this is the plan?)
2. We determine that we will have to keep using `unward` for the foreseeable. In that case, we will have to use it like `ward`, and thus need to restore the parts of the build system which transparently handled applying `ward` (now: `unward`) to sources followed by compilation (instead of working inlaces, the results would be placed into `gen/src/`; and build rules would ensure `gen/src/FOO.c` gets rebuilt whenever `src/FOO.c` or one of its header dependencies is touched.

Thing is, we need people to use this code so it gets tested, so keeping it in a PR indefinitely is not a good way forward.

Reply:
So, this is something that I don't have a good solution for myself. My plan was to not put the unward changes in for a transition period. The reason is that if we need to change something about how they work (e.g. include an option for disabling guards for performance reasons), they mess up diffs and git blame, as they're pretty extensive.

But that leads to the problem that if you build HPC-GAP without unward, you get a near immediate crash, as an unnecessary read or write guard fails (one that gets removed by unward).

Possible solution: add a `--without-guards` configure option. That allows builds to (unsafely) work without unward for the transition period *and* would also later allow HPC-GAP to be built without guards for performance reasons if performance is critical.

If we were to put unward into the build system, it already supports a `-o <dir>` option to place the transformed sources (and all untransformed ones, too) into a separate directory, say `gen/`. This takes less than a second, is incremental (it doesn't overwrite files that didn't get changed and therefore doesn't change their time stamps) and thus make-friendly, and is in principle possible.

But this would again introduce an external dependency, which I'm not very enthusiastic about (even though unward, unlike ward, is itself dependency-free and only requires a C++-98 or later compiler). A large part of the raison d'être for unward is, after all, that it can be dropped from the build system eventually and does not continue to be a maintenance burden.

Reply:
This sounds as if you are aiming for option 1 eventually. But the big question then is: when would we be ready for it? What's the hard criterium for making that decision? I keep coming back to this very same questions since when you first started talking about unward. You keep saying it's meant to be used once, but also keep saying that it may need to be changed, and then re-run. Unless there is a clear goal as to when it's "done", I don't see a solution out of this dilemma, other than to accept that the notion of "unward will be run once and then discarded" is hopeless. Hence option 2.

At the same time, it's also totally unclear to me what major changes to unward output you think might be needed?

Reply:
There are two possible changes that I'm contemplating right now and that may require changes in the output:

1. Allow disabling of guards at compile time.
2. Add `#ifdef HPCGAP` around the inline functions, so that they don't get compiled when not built with HPC-GAP. This was unnecessary when they were still static, but now they're not static anymore and should not show up in the code regardless.

My goal with unward was to keep the choice open-ended. This is why I specifically supported both the following options:

1. Make the process idempotent, so that you can rerun unward if necessary.
2. Add a `-o` option to allow for use as a preprocessor in the build system (i.e. like ward).

This means that we wouldn't be forced by design to pick one or the other.

As for the process, I was hoping to basically get the following done:

1. Add the current PR (cleaned up) to GAP.
2. Ask a couple of people to test that it works with unward (there aren't many who are using HPC-GAP, but there are a couple at least).
3. If that's the case, then after a short delay (couple of weeks) add the unward changes in a separate patch.

I am totally fine with rolling them into the current PR, too, if you think that's the best way forward. In the end, you probably have a better view of the big picture as far as GAP is concerned.

Reply:
@rbehrends (CC: @stevelinton @ChrisJefferson ) regarding your first two points, it seems you already added `--enable-guards` / `--disable-guards` (which is perfectly fine by me). Note sure around which inline functions you want(ed?) to add `#ifdef HPCGAP`, but I guess it's fine, too.


Regarding how to proceed: My main concern about merging this PR is how it would affect people who are using HPC-GAP right now, but who do not yet know about unward and thus are not using it. Will they get a horribly broken HPC-GAP, or just one that's more or less on the same level as what we have in `master` right now anyway?

If it was the latter, there is indeed nothing major stopping this from being merged (my remaining remarks are all very minor nitpicks and should be easy to resolve).

However, it seems to be the former (starting GAP built from your branch w/o running unward causes an immediate segfault due to violated guards). That makes me somewhat more reserved about the idea of merging this as-is. It'd be still OK by me if there was a very concrete timeline for it, and also a plan B should we determined that result of the merge is unusable. But my main concern is who the "couple of people" would be we'd ask to test it, and what makes us think they'd be more likely to test it then, than they would be to test this PR / your branch if we asked them explicitly to test it?

Until we come up with a good strategy to deal with that, it might indeed be best to split this commit into two: a non-controversial commit we could merge right away (which would contain the TLS changes, the renaming of `HAVE_NATIVE_TLS` to `USE_NATIVE_TLS` and a bunch of other minor changes), and the rest (mainly: the guards changes).